library/Faktory/Worker.hs

```diff
@@ -86,85 +96,76 @@ untilM_ predicate action = do
         untilM_ predicate action
     )
 
--- | Creates a new faktory worker, @'action'@ is ran with @'WorkerConfig'@ before
--- polling begins. Jobs received are passed to @'handler'@. The worker's
--- connection is closed when job processing ends.
-withRunWorker ::
-  (HasCallStack, FromJSON args)
-  =>  Settings
+-- | Forks a new faktory worker and continuously polls the faktory server for
+-- jobs which are passed to @'handler'@. The client is closed when the forked
+-- thread ends.
+startWorker
+  :: (HasCallStack, FromJSON args)
+  => Settings
   -> WorkerSettings
-  -> (WorkerConfig -> IO a)
   -> (Job args -> IO ())
-  -> IO ()
-withRunWorker settings workerSettings action handler =
-  configureWorker settings workerSettings
-    $ \config -> do
-        void $ action config
-        runWorkerWithConfig handler config
+  -> IO (Worker)
+startWorker settings workerSettings handler = do
+  wid <- maybe randomWorkerId pure $ settingsId workerSettings
+  isQuieted <- newTVarIO False
+  client <- newClient settings $ Just wid
+  isDone <- newEmptyMVar
+  let config = WorkerConfig{client, settings, wid, workerSettings}
+  tid <-
+    forkFinally
+      ( do
+          beatThreadId <- forkIOWithThrowToParent $ forever $ heartBeat config
+          finally
+            ( flip runReaderT config . runWorkerM $
+                catch
+                  (untilM_ (liftIO $ readTVarIO isQuieted) (processorLoop handler))
+                  (\(_ex :: WorkerHalt) -> pure ())
+            )
+            (killThread beatThreadId)
+      )
+      ( \e -> do
+          closeClient client
+          putMVar isDone ()
+          case e of
+            Left err ->
+              when
+                (isNothing $ fromException @WorkerHalt err)
+                (throwIO err)
```

Comment:
Your intent is for `err` to get rethrown to the consumer, right? Will this actually do that? Does the `finally` action in `forkFinally` run in the forked thread, or in the original thread? You might need to explicitly `throwTo` back to the parent, something like:

```hs
parentTid <- myThreadId
forkFinally
  (...)
  (\e -> throwTo parentTid e)
```

I suppose this isn't a super big deal, because exceptions thrown by the _job_ are always caught, so this would only be used for catching exceptions in the processor loop, which should never happen except in cases that would kill the whole process (e.g. OOM). I'll LGTM this, but if this might be worth investigating if you want to polish it a bit.